src/workerd/util/sqlite.c++

```diff
@@ -92,6 +92,58 @@ kj::String dbErrorMessage(int errorCode, sqlite3* db) {
   return msg.flatten();
 }
 
+// If a VFS call throws an exception, and vfsErrorListener is non-null, the exception will
+// be placed there, otherwise it will be logged. This is used to implement pass-through of KJ
+// exceptions through SQLite.
+static thread_local kj::Maybe<kj::Exception>* vfsErrorListener = nullptr;
+
+// Report that in a sqlite VFS callback, an exception was caught, and SQLITE_IOERROR is being
+// returned to SQLite.
+//
+// The exception must be caught using `catch (kj::Exception& e)`, NOT using `catch (...)` followed
+// by kj::getCaughtExceptionAsKj(). This is because the latter truncates the stack trace to show
+// only the frames between the throw and the catch. We actually want to retain the full trace
+// through SQLite.
+void reportVfsErrorCaught(kj::Exception&& e) {
+  if (vfsErrorListener != nullptr) {
+    // Only capture the first error; assume subsequent errors are side effects.
+    if (*vfsErrorListener == kj::none) {
+      *vfsErrorListener = kj::mv(e);
+    }
+  } else {
+    LOG_EXCEPTION("sqliteVfsError", e);
+  }
+}
+
+// Implements SQLITE_CALL_SCOPE.
+class SqliteCallScope {
+ public:
+  SqliteCallScope() {
+    KJ_DASSERT(vfsErrorListener == nullptr);
+    vfsErrorListener = &error;
+  }
+  ~SqliteCallScope() {
```

Comment:
Nit: It's likely going to be easy for someone to forget to call `rethrowVfsError`. Should the destructor do it for the user it is hasn't already been done? Or maybe emit a warning if `error` is not `kj::none` and rethrow hasn't been called?